examples/generative/ddpm.py

```diff
@@ -0,0 +1,682 @@
+    def apply(inputs):
+        batch_size = tf.shape(inputs)[0]
+        height = tf.shape(inputs)[1]
+        width = tf.shape(inputs)[2]
+        scale = tf.cast(units, tf.float32) ** (-0.5)
+
+        inputs = layers.GroupNormalization(groups=groups)(inputs)
+        q = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+        k = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+        v = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+
+        attn_score = tf.einsum("bhwc, bHWc->bhwHW", q, k) * scale
+        attn_score = tf.reshape(attn_score, [batch_size, height, width, height * width])
+
+        attn_score = tf.nn.softmax(attn_score, -1)
+        attn_score = tf.reshape(attn_score, [batch_size, height, width, height, width])
+
+        proj = tf.einsum("bhwHW,bHWc->bhwc", attn_score, v)
+        proj = layers.Dense(units, kernel_initializer=kernel_init(0.0))(inputs)
+        return inputs + proj
+
+    return apply
+
+
+def TimeEmbedding(inputs, dim=32):
+    """Transformer sinusoidal position embedding."""
+    half_dim = dim // 2
+    emb = math.log(10000) / (half_dim - 1)
+    emb = tf.exp(tf.range(half_dim, dtype=tf.float32) * -emb)
+
+    inputs = tf.cast(inputs, dtype=tf.float32)
+    emb = inputs[:, None] * emb[None, :]
+    emb = tf.concat([tf.sin(emb), tf.cos(emb)], axis=-1)
+    return emb
+
+
+def ResidualBlock(width, groups=8):
+    def apply(inputs):
+        x, t = inputs
+        input_width = x.shape[3]
+
+        if input_width == width:
+            residual = x
+        else:
+            residual = layers.Conv2D(
+                width, kernel_size=1, kernel_initializer=kernel_init(1.0)
+            )(x)
+
+        temb = act_fn(t)
+        temb = layers.Dense(width, kernel_initializer=kernel_init(1.0))(temb)[
+            :, None, None, :
+        ]
+
+        x = layers.GroupNormalization(groups=groups)(x)
+        x = act_fn(x)
+        x = layers.Conv2D(
+            width, kernel_size=3, padding="same", kernel_initializer=kernel_init(1.0)
+        )(x)
+
+        x = layers.Add()([x, temb])
+        x = layers.GroupNormalization(groups=groups)(x)
+        x = act_fn(x)
+
+        x = layers.Conv2D(
+            width, kernel_size=3, padding="same", kernel_initializer=kernel_init(0.0)
+        )(x)
+        x = layers.Add()([x, residual])
+        return x
+
+    return apply
+
+
+def DownSample(width):
+    def apply(x):
+        x = layers.Conv2D(
+            width,
+            kernel_size=3,
+            strides=2,
+            padding="same",
+            kernel_initializer=kernel_init(1.0),
+        )(x)
+        return x
+
+    return apply
+
+
+def UpSample(width):
+    def apply(x):
+        x = layers.UpSampling2D(size=2, interpolation="nearest")(x)
+        x = layers.Conv2D(
+            width, kernel_size=3, padding="same", kernel_initializer=kernel_init(1.0)
+        )(x)
+        return x
+
+    return apply
+
+
+def TimeMLP(temb, units):
```

Comment:
Why not use the `apply()` pattern like above?

Reply:
Yes, I will do that